File: microservices/boxes/boxes/Models/SupplierCategory.js
const mongoose = require("mongoose");
const Schema = mongoose.Schema;
const slugify = require('slugify');
const ObjectId = Schema.Types.ObjectId

const SupplierCategorySchema = Schema({
    sort: {type: Number, required: false, default: 0},
    tenant_id: {type: String, required: true},
    tenant_name: {type: String, required: true},
    countries: {type: Array, required: false, default: []},
    sku: {type: String, required: false, default: ''},
    name: {type: String, required: true},
    system_key: {type: String, required: true},
    display_name: {type: Array, required: true},
    slug: {type: String, required: false},
    description: {type: String, required: false, default: ''},
    shareable: {type: Boolean, required: false, default: false},
    published: {type: Boolean, required: false, default: true},
    media: {type: Array, required: false, default: []},
    price_build: {type: Object, required: false, default: {}},
    has_products: {type: Boolean, required: false, default: false},
    has_manifest: {type: Boolean, required: false, default: false},
    calculation_method: {type: Array, required: false, default: []},
    dlv_days: {type: Array, required: false, default: []},
    printing_method: {type: Array, required: false, default: []},
    production_days: {type: Array, required: false, default: []},
    production_dlv: {type: Array, required: false, default: []},
    start_cost: {type: Number, required: false, default: 0},
    linked: { type: ObjectId, required: false, index: true , default: null},
    ref_id: {type: String, required: false, default: ''},
    ref_category_name: {type: String, required: false, default: ''},
    bleed: {type: Number, required: false, default: 0},
    range_list: {type: Array, required: false, default: []},
    ranges: {type: Array, required: false, default: []},
    limits: {type: Array, required: false, default: []},
    free_entry: {type: Array, required: false, default: []},
    range_around: {type: Number, required: false, default: 0},
    vat: {type: Number, required: false, default: 0},
    additional: {type: Array, required: false, default: []},
    created_at: {type: Date, default: Date.now, index: true},
    updated_at: {type: Date, default: Date.now, index: true},
}, {
    versionKey: false // You should be aware of the outcome after set to false
});


// Pre-save hook to dynamically set the sort value and slug
SupplierCategorySchema.pre('save', async function (next) {
    try {

        // Automatically set the slug if the name field is modified
        if (this.isModified('name')) {
            this.slug = slugify(this.name, { lower: true });
        }

        next();
    } catch (err) {
        next(err);
    }
});
module.exports = SupplierCategory = mongoose.model("supplier_categories", SupplierCategorySchema);

File: microservices/old-finder/models/supplierCategory.py
from database.db import db
from flask_mongoengine import BaseQuerySet
from slugify import slugify, Slugify, UniqueSlugify
from models.category import Category
import uuid
import datetime


def id(): return str(uuid.uuid4())


class SupplierCategory(db.Document):
    sort = db.IntField(required=False, default=0)
    tenant_id = db.StringField(required=True)
    tenant_name = db.StringField(required=True)
    countries = db.ListField(required=False, default=[])
    sku = db.StringField(required=False, default='')
    name = db.StringField(required=True)
    display_name = db.ListField(required=True, default=[])
    system_key = db.StringField(required=True)
    slug = db.StringField(required=True, default='', unique_with=('tenant_id'))
    description = db.StringField(required=False, default=None)
    shareable = db.BooleanField(required=False, default=False)
    published = db.BooleanField(required=False, default=True)
    media = db.ListField(required=False)
    price_build = db.DictField(required=False, default={})
    has_products = db.BooleanField(required=False, default=False)
    has_manifest = db.BooleanField(required=False, default=False)
    calculation_method = db.ListField(required=False,)
    production_days = db.ListField(required=False, default=[])
    dlv_days = db.ListField(required=False, default=[])
    printing_method = db.ListField(required=False, default=None)
    start_cost = db.IntField(required=False, default=0)
    additional = db.DictField(required=False, default={})
    ref_id = db.StringField(required=False, default='')
    ref_category_id = db.ReferenceField('self', required=False)
    ref_category_name = db.StringField(required=False, default='')
    linked = db.ReferenceField(Category)
    created_at = db.DateTimeField(default=datetime.datetime.now, required=True)

    meta = {
        'indexes': ['-created_at', 'slug', 'name', 'linked'],
        'ordering': ['-created_at'],
        'collection': 'supplier_categories',
        'queryset_class': BaseQuerySet
    }

    def __setattr__(self, key, value):
        super(SupplierCategory, self).__setattr__(key, value)
        if key == 'name':
            self.slug = slugify(self.name, to_lower=True)

